### PR TITLE
Fix Arcade for ASPNETCore

### PR DESCRIPTION
**Pull request by missymessa**
Related: https://github.com/dotnet/core-eng/issues/11779

The new base class to support DI for MSBuild Tasks in Arcade needed to support a workaround for an issue with binding redirects. After providing that fix, discovered that we needed to also downgrade the version of Microsoft.Extensions.DependencyInjection library (and related libraries) to 2.1.1.

### CODE REVIEW

File: src/Microsoft.Arcade.Test.Common/Microsoft.Arcade.Test.Common.csproj [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp2.1</TargetFramework>
```
Comment (chcosta) [inline src/Microsoft.Arcade.Test.Common/Microsoft.Arcade.Test.Common.csproj:4]: There are tests that run on net472, are they not supported via this model at this time?

File: src/Microsoft.Arcade.Test.Common/Microsoft.Arcade.Test.Common.csproj [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Build" Version="$(MicrosoftBuildVersion)" />
```
Comment (chcosta) [inline src/Microsoft.Arcade.Test.Common/Microsoft.Arcade.Test.Common.csproj:8]: It looks like every project that includes Microsoft.Arcade.Test.Common, also includes Microsoft.Arcade.Common, can we just merge the two into Microsoft.Arcade.Common and conditionalize the packagereference on "IsTestProject"?

Comment (missymessa) [reply src/Microsoft.Arcade.Test.Common/Microsoft.Arcade.Test.Common.csproj:8]: I think it's better to not couple these projects together. Not every project that references Microsoft.Arcade.Common also references Microsoft.Arcade.Test.Common.

File: src/Microsoft.Arcade.Common/src/Microsoft.Arcade.Common/MSBuildTaskBase.Desktop.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
[the remark this answers is not in the document]
Comment (missymessa) [reply src/Microsoft.Arcade.Common/src/Microsoft.Arcade.Common/MSBuildTaskBase.Desktop.cs]: Moved these files.

File: src/Microsoft.Arcade.Common/src/Microsoft.Arcade.Common/MSBuildTaskBase.cs [hunk 4]
```diff
@@ -29,10 +29,13 @@ public abstract class MSBuildTaskBase : MSBuild.Task
...
-            ServiceCollection collection = new();
+            ServiceCollection collection = new ServiceCollection();
             ConfigureServices(collection);
-            using var provider = collection.BuildServiceProvider();
-            return InvokeExecute(provider);
+            using (var provider = collection.BuildServiceProvider())
+            {
+                return InvokeExecute(provider);
+            }
+
```
[the remark this answers is not in the document]
Comment (missymessa) [reply src/Microsoft.Arcade.Common/src/Microsoft.Arcade.Common/MSBuildTaskBase.cs]: Removed.

File: src/Microsoft.Arcade.Test.Common/Microsoft.Arcade.Test.Common.csproj (hunk 1 above)
Comment (missymessa) [reply src/Microsoft.Arcade.Test.Common/Microsoft.Arcade.Test.Common.csproj]: Added the net472 TFM to this list so that it can be used by net472 projects in the future.

File: eng/Versions.props [hunk 5]
```diff
@@ -30,9 +30,10 @@
     <MicrosoftIdentityModelClientsActiveDirectoryVersion>3.17.2</MicrosoftIdentityModelClientsActiveDirectoryVersion>
     <MicrosoftRestClientRuntimeVersion>2.3.13</MicrosoftRestClientRuntimeVersion>
     <MicrosoftExtensionsDependencyModelVersion>2.1.0</MicrosoftExtensionsDependencyModelVersion>
-    <MicrosoftExtensionsDependencyInjectionVersion>3.1.9</MicrosoftExtensionsDependencyInjectionVersion>
+    <MicrosoftExtensionsDependencyInjectionAbstractionsVersion>2.1.1</MicrosoftExtensionsDependencyInjectionAbstractionsVersion>
+    <MicrosoftExtensionsDependencyInjectionVersion>2.1.1</MicrosoftExtensionsDependencyInjectionVersion>
```
Comment (ericstj) [inline eng/Versions.props:34]: Why did you need to downgrade?  Were there other assemblies in the same process using older version?  Were there breaking changes that prevented you from redirecting to the new version?  cc @eerhardt @maryamariyan

Comment (missymessa) [reply eng/Versions.props:34]: After I added the change you suggested in, I got a new error :)  

Link to build with error: https://dnceng.visualstudio.com/internal/_build/results?buildId=937937&view=logs&j=1b89928a-2219-5ef9-602f-f95beb3da4dc&t=b8e57a46-4a1c-5752-600e-7a395fe90fb9&l=234

```
F:\workspace\_work\1\s\.packages\microsoft.dotnet.arcade.sdk\6.0.0-beta.21054.5\tools\Publish.proj(151,5): error MSB4018: The "PushToAzureDevOpsArtifacts" task failed unexpectedly.
F:\workspace\_work\1\s\.packages\microsoft.dotnet.arcade.sdk\6.0.0-beta.21054.5\tools\Publish.proj(151,5): error MSB4018: System.TypeLoadException: Method 'DisposeAsync' in type 'Microsoft.Extensions.DependencyInjection.ServiceProvider' from assembly 'Microsoft.Extensions.DependencyInjection, Version=3.1.9.0, Culture=neutral, PublicKeyToken=adb9793829ddae60' does not have an implementation.
F:\workspace\_work\1\s\.packages\microsoft.dotnet.arcade.sdk\6.0.0-beta.21054.5\tools\Publish.proj(151,5): error MSB4018:    at Microsoft.Arcade.Common.MSBuildTaskBase.Execute()
F:\workspace\_work\1\s\.packages\microsoft.dotnet.arcade.sdk\6.0.0-beta.21054.5\tools\Publish.proj(151,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
F:\workspace\_work\1\s\.packages\microsoft.dotnet.arcade.sdk\6.0.0-beta.21054.5\tools\Publish.proj(151,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()
##[error].packages\microsoft.dotnet.arcade.sdk\6.0.0-beta.21054.5\tools\Publish.proj(151,5): error MSB4018: (NETCORE_ENGINEERING_TELEMETRY=Publish) The "PushToAzureDevOpsArtifacts" task failed unexpectedly.
System.TypeLoadException: Method 'DisposeAsync' in type 'Microsoft.Extensions.DependencyInjection.ServiceProvider' from assembly 'Microsoft.Extensions.DependencyInjection, Version=3.1.9.0, Culture=neutral, PublicKeyToken=adb9793829ddae60' does not have an implementation.
   at Microsoft.Arcade.Common.MSBuildTaskBase.Execute()
   at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
   at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()

Build FAILED.
```

Originally, I had thought that it had to do with the single-line using statement I had in there, but even when I reverted to the old-style of `using(){ }`, it still failed. The projects in Arcade also target .NET Core 2.1, and subsequently, downgrading the packages to 2.1 (2.1.1, specifically) fixed the error: 

Link to build that succeeded: https://dnceng.visualstudio.com/internal/_build/results?buildId=938715&view=results

If there is a better way of resolving this without the downgrade, I'd be happy to implement that instead.

**Review by ChadNedzlek — APPROVED** [review]
It's weird that stuff is still using dotnet 2.1 to build things, since we are working on .NET 6 at this point. I wonder if we should open an issue to upgrade whatever is so ancient somehow?

File: src/Microsoft.Arcade.Common/Microsoft.Arcade.Common.csproj [hunk 6]
```diff
@@ -1,13 +1,20 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFrameworks>netcoreapp2.1;net472</TargetFrameworks>
-    <IsPackable>true</IsPackable>
```
Comment (alexperovich) [inline src/Microsoft.Arcade.Common/Microsoft.Arcade.Common.csproj:5]: Why remove the "IsPackable"?

File: src/Microsoft.Arcade.Common/Microsoft.Arcade.Common.csproj [hunk 7]
```diff
@@ -1,13 +1,20 @@
...
   <ItemGroup>
     <PackageReference Include="Microsoft.Build.Utilities.Core" Version="$(MicrosoftBuildUtilitiesCoreVersion)" />
     <PackageReference Include="Microsoft.Extensions.DependencyInjection" Version="$(MicrosoftExtensionsDependencyInjectionVersion)" />
   </ItemGroup>
 
+  <ItemGroup Condition="'$(TargetFramework)' == 'net472'">
+    <Compile Include="..\Common\AssemblyResolver.cs" />
+  </ItemGroup>
+  
+  <ItemGroup Condition="'$(TargetFramework)' != 'net472'">
```
Comment (alexperovich) [inline src/Microsoft.Arcade.Common/Microsoft.Arcade.Common.csproj:16]: This condition looks backwards. This item should be included when targeting net472

File: src/Microsoft.Arcade.Common/Microsoft.Arcade.Common.csproj (hunk 6 above)
Comment (missymessa) [reply src/Microsoft.Arcade.Common/Microsoft.Arcade.Common.csproj:5]: I didn't think it needed it? Should I leave it in?

File: src/Microsoft.Arcade.Common/Microsoft.Arcade.Common.csproj (hunk 7 above)
Comment (missymessa) [reply src/Microsoft.Arcade.Common/Microsoft.Arcade.Common.csproj:16]: Since the file lives in the same folder with the rest of the files in the project, VS is automatically adding it as a part of the project. I'd have to add the exclusion earlier and then bring it in with the conditional afterwards, which seems just as weird to me.

File: src/Microsoft.Arcade.Common/Microsoft.Arcade.Common.csproj (hunk 6 above)
Comment (alexperovich) [reply src/Microsoft.Arcade.Common/Microsoft.Arcade.Common.csproj:5]: its required if you want this project to make a nupkg

File: src/Microsoft.Arcade.Common/Microsoft.Arcade.Common.csproj (hunk 7 above)
Comment (alexperovich) [reply src/Microsoft.Arcade.Common/Microsoft.Arcade.Common.csproj:16]: O wait, I noticed this is an exclusion. never mind.

File: src/Microsoft.Arcade.Common/Microsoft.Arcade.Common.csproj (hunk 6 above)
Comment (missymessa) [reply src/Microsoft.Arcade.Common/Microsoft.Arcade.Common.csproj:5]: kk, I'll add it back in.

**Review by alexperovich — APPROVED** [review]
The downgrade is concerning, but otherwise LGTM.

**Comment by missymessa** [comment]
> The downgrade is concerning, but otherwise LGTM.

Agreed. I hope we can find another solutions so that we don't have to do that.

File: eng/Versions.props (hunk 5 above)
Comment (ChadNedzlek) [reply eng/Versions.props:34]: It looks like the netstanard2.0 package (which presumably we are getting if we compile for netcoreapp2.1) implements a shimmed IAsyncDisposable from Microsoft.Bcl.AsyncInterfaces.  And then this code is looking for the "real" IAsyncDisposable, and can't find it.  It seems there is a path through the DI libraries that just blows up in your face.  If you sort of go from netstandard2.0 and end up executing in a netcore3.1 universe, you look for the wrong IAsyncDisposable.

Downgrading helps because it's before all the IAsyncDispose rigamarole, so you don't have to worry about it. :-)

Comment (ericstj) [reply eng/Versions.props:34]: BCL.AsyncInterfaces should contain multiple implementations, with appropriate typeforwards when the type is inbox.  If you're running on .NETCore 3.1 you should only be loading the BCL.AsyncInterfaces that has typeforwards if you need it to exchange with other code that's compiled against 3.1.  This should happen when you publish your task for 3.1.

> you sort of go from netstandard2.0 and end up executing in a netcore3.1 universe, you look for the wrong IAsyncDisposable.

Somewhere in your layout you have something mixing code which is compiling against 3.1 with code that published AsyncInterfaces for a different framework (netstandard2.0, or older netcoreapp).  Can you point me to to the package that reproduces this problem?  I might be able to spot the culprit.

Comment (missymessa) [reply eng/Versions.props:34]: The `Microsoft.Extensions.DependencyInjection` package is what I downgraded. It gets used in a few projects in Arcade, now. For this feature, it was added to the Microsoft.DotNet.Build.Tasks.Feed project and a new project called Microsoft.Arcade.Common. These projects target `net472` and `netcoreapp2.1`

Comment (ericstj) [reply eng/Versions.props:34]: Well if everything were binding against the netcoreapp2.1 implementation of AsyncInterfaces, no-one would ever see/care about type conflicts with IAsyncDisposable because everyone would be agreeing that it lived in Microsoft.BCL.AsyncInterfaces.  What your callstack indicated is that someone thinks IAsyncDisposable lives in a different assembly.  It seems to be this:
https://github.com/dotnet/arcade/blob/6261cf2e7411aa1e369e4f5c019419b4704f59ec/src/Microsoft.Arcade.Common/src/Microsoft.Arcade.Common/MSBuildTaskBase.cs#L34
Since your code is targeting netcoreapp2.1 it thinks IAsyncDisposable lives in MS.BCL.AsyncInterfaces.  

I wonder if someone else running in the same process built and published for netcoreapp3.1 and loaded the extensions assemblies before you could.  That might explain it.  By downgrading you are just avoiding assembly reuse.  Why not just retarget to 3.1?  Or include a 3.1 build in the package.

Comment (missymessa) [reply eng/Versions.props:34]: Let me run a build targeting 3.1 to see if there are any issues with that.

Comment (ericstj) [reply eng/Versions.props:34]: I see, most of arcade is still targeting .NETCoreApp2.1.  That's odd.  We should probably update that and it would fix this problem.  You could fork that effort off into a separate PR if you've already validated this.  You could undo the downgrade when retargeting.

Comment (missymessa) [reply eng/Versions.props:34]: I just attempted to update only the select projects that are using the new abstract class (or referencing the Microsoft.Extensions.DependencyInjection library) to get DI in the MSBuild Tasks and got that same error about DisposeAsync again. I think we're going to need a larger overhaul to upgrade Arcade to 3.1 (or even better .NET 5), but that should be in a different PR. I think I'll merge this to get Arcade unblocked for now and write up an issue to do a more complete framework upgrade.

Comment (missymessa) [reply eng/Versions.props:34]: Created this to handle upgrading the entire repo to at least 3.1, ideally 5: https://github.com/dotnet/arcade/issues/6745

**Comment by missymessa** [comment]
For more visibility, created https://github.com/dotnet/arcade/issues/6745 to handle upgrading the .NET Core framework.
